migratekvdb/migration.go

```diff
@@ -0,0 +1,1140 @@
+)
+
+// Config holds the configuration for the migrator.
+type Config struct {
+	// ChunkSize is the number of items (key-value pairs or buckets) to
+	// process in a single transaction in bytes.
+	ChunkSize uint64
+
+	// Logger is the logger to use for logging.
+	Logger btclog.Logger
+
+	// MetaDB is the database that stores the migration/verification state.
+	// We store it separately to not clutter the source or destination
+	// databases.
+	MetaDB *bbolt.DB
+}
+
+// validateConfig ensures the configuration is valid and sets defaults.
+func validateConfig(cfg *Config) error {
+	if cfg == nil {
+		return fmt.Errorf("config cannot be nil")
+	}
+
+	// Validate and set defaults for chunk size.
+	switch {
+	case cfg.ChunkSize == 0:
+		cfg.ChunkSize = DefaultChunkSize
+
+		cfg.Logger.Infof("Chunk size set to default: %d bytes",
+			cfg.ChunkSize)
+
+	case cfg.ChunkSize > MaxChunkSize:
+		return fmt.Errorf("chunk size too large: %d bytes, maximum "+
+			"is %d bytes", cfg.ChunkSize, MaxChunkSize)
+	}
+
+	// Ensure logger is set.
+	if cfg.Logger == nil {
+		cfg.Logger = btclog.Disabled
+	}
+
+	return nil
+}
+
+// Migrator handles the chunked migration of bolt databases. It supports:
+//   - Resumable migrations through state tracking
+//   - Chunk-based processing to handle large databases
+//   - Verification of migrated data (also resumable)
+//   - Progress logging
+type Migrator struct {
+	// cfg is the configuration for the migrator.
+	cfg Config
+
+	// migration is the migration state which is used to track and also
+	// resume the migration.
+	migration MigrationState
+
+	// verification is the verification state which is used to track and
+	// also resume the verification.
+	verification VerificationState
+
+	// Rate tracking fields to track the progress of the migration and
+	// verification.
+	startTimeMigration     time.Time
+	startTimeVerification  time.Time
+	lastLogTime            time.Time
+	logInterval            time.Duration
+	lastMigratedKeyCount   int64
+	lastVerifiedKeyCount   int64
+	migratedKeysSinceStart int64
+	verifiedKeysSinceStart int64
+}
+
+// New creates a new Migrator with the given configuration.
+func New(cfg Config) (*Migrator, error) {
+	// Validate and set defaults for the config.
+	if err := validateConfig(&cfg); err != nil {
+		return nil, fmt.Errorf("invalid config: %w", err)
+	}
+
+	return &Migrator{
+		cfg:                cfg,
+		startTimeMigration: time.Now(),
+		lastLogTime:        time.Now(),
+		logInterval:        5 * time.Second,
+	}, nil
+}
+
+// markMigrationComplete marks the migration as complete in the migration_meta
+// bucket.
+func (m *Migrator) markMigrationComplete() error {
+	return m.cfg.MetaDB.Update(func(tx *bbolt.Tx) error {
+		metaBucket, err := tx.CreateBucketIfNotExists(
+			[]byte(migrationMetaBucket),
+		)
+		if err != nil {
+			return fmt.Errorf("failed to create meta "+
+				"bucket: %w", err)
+		}
+
+		timestamp := time.Now().UTC().Format(time.RFC3339)
```

Comment:
nit: since this is repeated pattern, perhaps worth adding a helper `getTimestamp()` fn.